whist/core/cards/card_container.py

```diff
@@ -1,60 +1,224 @@
+    cards: tuple[Card, ...]
+
+    # pylint: disable=too-few-public-methods
+    class Config:
+        """
+        Configuration class for base model to make it immutable.
+        """
+        allow_mutation = False
+
+    @classmethod
+    def empty(cls) -> 'CardContainer':
+        """
+        Creates an empty card container.
+
+        :return: empty ard container
+        :rtype: correct subtype of CardContainer
+        """
+        return cls(cards=())
+
+    @classmethod
+    def with_cards(cls, *cards) -> 'CardContainer':
+        """
+        Creates a card container with the given cards.
+
+        :param cards: cards to add
+        :return: card container with given cards
+        :rtype: correct subtype of CardContainer
+        """
+        if len(cards) == 1 and not isinstance(cards[0], Card):
+            cards = cards[0]
+        return cls(cards=cards)
+
+    @classmethod
+    def full(cls) -> 'CardContainer':
+        """
+        Create a full card container.
+
+        :return: full card container
+        :rtype: correct subtype of CardContainer
         """
-        Constructor
+        return cls(cards=Card.all_cards())
 
-        :param args: multiple cards or one card iterable
+    def pop_random(self) -> Card:
         """
-        if len(args) == 1 and not isinstance(args[0], Card):
-            self._cards = {*args[0]}
-        else:
-            self._cards = {*args}
+        Removes one random card from card container.
+
+        :return: A card from deck.
+        """
+        card = random.choice(self.cards)
+        self.remove(card)
+        return card
 
     def __contains__(self, card: Card) -> bool:
-        return card in self._cards
+        return card in self.cards
 
     def __len__(self):
-        return len(self._cards)
+        return len(self.cards)
 
     def __iter__(self) -> Iterator[Card]:
-        return iter(self._cards)
+        return iter(self.cards)
 
     def __str__(self) -> str:
-        return str(self._cards)
+        return str(self.cards)
 
     def __repr__(self) -> str:
-        return f'CardContainer(cards={self._cards!r})'
-
-    def __eq__(self, other: Any) -> bool:
-        if self.__class__ is other.__class__:
-            # pylint: disable=protected-access
-            return self._cards == other._cards
-        return NotImplemented
+        return f'{self.__class__.__name__}(cards={self.cards!r})'
 
     def remove(self, card: Card) -> None:
         """
         Remove a card from this container.
 
         :param card: card to remove
         """
-        self._cards.remove(card)
+        if not isinstance(card, Card):
+            raise ValueError(f'cannot remove {card} of type {type(card)} from card container')
+        if card not in self:
+            raise ValueError(f'{card} not in card container')
+        self._remove_impl(card)
+
+    @abc.abstractmethod
+    def _remove_impl(self, card: Card) -> None:
+        pass
```

Comment:
Shouldn't we raise `NotImplementedError` here?

Reply:
it does not matter. this method does not really exist and will never be called. i will change it